### PR TITLE
*DRAFT* Firmware updates

### PR DESCRIPTION
**Pull request by lulf**
This is an initial draft describing the high-level concept of firmware updates covering drogue-device and drogue-cloud. Before extending this proposal with the specific APIs (any help with that would be appreciated) and functionality of drouge-device, I'd like us to reach concensus on the high level design.

Readable version: https://github.com/drogue-iot/rfcs/blob/feature/firmware-updates/active/0006-firmware-updates.md

### CODE REVIEW

**Review by dejanb — APPROVED** [review]
I like the general concept of DFU Controller and using standard commands to communicate with the device.

**Comment by jbtrystram** [comment]
Awesome @lulf ! Reading it felt like the early stages of my PhD :) 
If I understand correctly the bootloader here act as a "static" part, a bit like the UEFI firmware on a PC motherboard. 
The tricky question is: how do you update the bootloader ? If I keep the computer analogy it can be updated from code running in the memory. But in the case of a remote IoT device it is hard to recover if something goes wrong.

**Comment by lulf** [comment]
> If I understand correctly the bootloader here act as a "static" part, a bit like the UEFI firmware on a PC motherboard.
> The tricky question is: how do you update the bootloader ? If I keep the computer analogy it can be updated from code running in the memory. But in the case of a remote IoT device it is hard to recover if something goes wrong.
> 

@jbtrystram That is a good point. My initial opinion is that we don't support updating the bootloader without physical access to begin with. During development of the bootloader (or re-use some existing available), we might figure out how complex it is and if we think we are going to need that functionality and plan for it. I don't think you can get around with a minimal amount of code that must _always work_ (a boot-bootloader if you will) unless there is hardware support for having multiple alternating entry points or something. So the tradeoff is how complex the bootloader vs. flash usage.

My initial thought is that the bootloader should only have the ability to load one out of two firmwares, a way to write flash, and switch between the two versions of the firmware.

> If my memory serves corectly one solution was to have double the space needed for the bootloader so you could write the new bootloader aside to the old one then verify it so and after all have been checked you could erase the old one. Using double the flash space is a big compromise though.

In the initial proposal, I am suggesting we use half of the space of the _firmware_ itself, in order to be able to load an earlier version of the firmware and switch between them. This is to avoid the bootloader needing network support in order to flash an older version of the firmware. 

The argument for not having network capabilities in the bootloader is that we don't really know the network capabilities of the devices we are going to support (it could be one out of many different drivers). It also reduces the chance that we need to update the bootloader itself. 

So, the implication here is that `firmware size < flash size /2`.

I have updated the bootloader detail section with some ideas.

**Comment by jbtrystram** [comment]
Thanks for expanding on that, I think the assumptions you make are reasonable. To me limiting the firmware to half the flash space feels like a huge limitation but i too don't have enough experience to back it up. 
I thought that the bootloader would have the network capabilities, but now i understand better. I am all for the smallest amount of bootloader code as possible.

**Comment by lulf** [comment]
An alternative approach that could save some flash would be to optionally bundle the bootloader with the minimal code to do networking: actor framework and a selected network driver configured at build time + DFU actor. In theory, that should consume less flash space than the full application. However, that 'minimal code' would need to be updateable.

File: active/0006-firmware-updates.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+* (User facing) Publish a new firmware version
+* (User facing) Update firmware for a group of devices
+* (Device facing) Poll/Notify device about a new firmware version
+* (Device facing) Fetch the new firmware
+
+Drogue-device needs a bootloader that can:
+
+* Boot firmwares (duh)
+* Write new firmware versions to flash while current firmware is running (banking)
```
Comment (bobmcwhirter) [inline active/0006-firmware-updates.md:35]: I'm of the mind that a bootloader should be able to boot from banks, possibly copying from long-term storage to an executable bank of memory (maybe?), but populating the bank is not directly the bootloader's job. That being said, the bootloader may include a userland library/api compiled with the app to assist in bank management.

File: active/0006-firmware-updates.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+A new component in drogue cloud, `DFU Controller`, is responsible for polling Hawkbit to check if there is a new firmware available for devices, and initiating the firmware update. When a new firmware is availble, the `DFU Controller` will send a command to the device asking it to update its firmware. The existing command API is used to notify devices of new firmware versions, using a predefined payload that is understood by drogue-device.
+
+When the device has received the command to update its firmware, it will start the process of fetching the firmware and writing it to flash. Updating the firmware of the device must happen in chunks, as the firmware size is usually larger than the amount of RAM on the device. This chunking process is handled by passing the firmware size in the DFU command.
+
+Once the firmware update is complete, the device will update the bootloader to use the new firmware and reset itself. During the update, and after reset, the device may send feedback to the HTTP endpoint of its progress and status.
```
Comment (bobmcwhirter) [inline active/0006-firmware-updates.md:62]: Agreed. I could also see a strong division between "fetching firmware" and "reboot with it".

For instance, perhaps a fleet has firmware rolled out to it, inactive, for a month, until all devices have it onboard, but not running. Subsequently, "boot into firmware N+1" command is separately issued causing remote reboots/upgrades, possibly in a rolling/canary fashion.

File: active/0006-firmware-updates.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+* TODO: DFU command spec
+
+## Drogue Device
+
+### Bootloader
+
+* To simplify the design, the bootloader cannot be updated without re-flashing the device. 
+* The bootloader will keep a 'current' firmware that it will boot first.
+* If the device panics, it will retry the 'current' firmware until a max number of retries.
+* If it panics and reaches max number of retries, diagnostics will be stored, and the 'previous' firmware will be used.
```
Comment (bobmcwhirter) [inline active/0006-firmware-updates.md:84]: Let's also investigate mcuBoot. I don't think we want to directly use it, but it's got concepts of partition tables (slots/banks), verifying signatures/integrity prior to booting, and post-boot validation before committing to the new variant.

Also, I think there may be support for "factory" reset using a 3rd permanent bank (possibly on long-term storage) for button-based local hard-reset to known good state.

File: active/0006-firmware-updates.md (hunk 2 above)
Comment (lulf) [reply active/0006-firmware-updates.md:62]: Good point, I will clarify that distinction.

File: active/0006-firmware-updates.md (hunk 3 above)
Comment (lulf) [reply active/0006-firmware-updates.md:84]: Agreed on all points, we should consider different alternatives. I like the factory reset idea. The challenge I think is going to be the difference capabilities in boards we are going to support.

File: active/0006-firmware-updates.md (hunk 1 above)
Comment (lulf) [reply active/0006-firmware-updates.md:35]: I don't know enough about how different boards support this, but it needs to be investigated so we can make a comparison and find a common denominator and/or how to expose the differences.

File: active/0006-firmware-updates.md (hunk 3 above)
Comment (bobmcwhirter) [reply active/0006-firmware-updates.md:84]: Right on, I think flexibility will be key, and viewing it as a bootloader framework to be applied as needed by the end-user. Some super-constrained devices that don't expect to update might only have 1 partition or not check signatures, or not have a factory-reset from long-term storage.

**Pull request merged**